### PR TITLE
Series and Series expansion.

### PR DESCRIPTION
**Pull request by shivamvats**
@leosartaj I have reopened our PR here. Were you able to compile the code you added?

### CODE REVIEW

**Comment by shivamvats** [comment]
@leosartaj I have added you as a collaborator.

**Comment by asmeurer** [comment]
You also could have created a branch on this repo.

File: features.tex [hunk 1]
```diff
@@ -17,6 +17,10 @@ \subsection{Matrices}
 % Physics module (some sampling, to show that it is there)
 \subsection{Physics}
 
+% Series module (Formal Power Series, Fourier Series)
+\subsection{Series}
+\input{series.tex}
```
Comment (leosartaj) [inline features.tex]: `\input{series}` ?

Comment (shivamvats) [inline features.tex]: This imports content of another .tex file. This way fewer merge conflicts for everyone.

File: series.tex [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+>>> x, y = symbols('x, y')
+>>> series(sin(x + y) + cos(x*y), x, 0, 2)
+1 + sin(y) + x*cos(y) + O(x**2)
+\end{verbatim}
+
+The newer and much faster approach called Ring Series makes use of the
+observation that a truncated Taylor series, is in fact a polynomial.
+Ring-series uses the efficient representation and operations of sparse
+polynomials. The choice of sparse polynomials is deliberate as it performs
+well in a wider range of cases than a dense representation. It gives the user
```
Comment (isuruf) [inline series.tex]: It -> Ring series

File: series.tex [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+the freedom to choose the type of coefficients he wants to have in his series,
+allowing the use of faster operations on certain types.
+
+For this, several low level methods for expansion of elementary functions are
+implemented using semi-numerical algorithms. All these support Puiseux seris
```
Comment (isuruf) [inline series.tex]: seris -> series

File: series.tex [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+>>> from sympy.polys.ring_series import rs_sin
+>>> R, x = ring('x', QQ)
+>>> rs_sin(x**2 + x, x, 5)
+-1/2*x**4 - 1/6*x**3 + x**2 + x
+\end{verbatim}
+
+The function \texttt{rs\_series} makes use of these elementary functions to
+intelligently expand an arbitrary SymPy expression. Currently it supports
+expansion only about 0 and is under active development. Typical speedup over
+the older SymPy implementation is in the range 20-100 times with larger
```
Comment (isuruf) [inline series.tex]: Value 20-100 is given with no benchmarks.

Comment (leosartaj) [inline series.tex]: We should add a citation for the benchmark. Ping @shivamvats

**Comment by leosartaj** [comment]
Please review. Ping @isuruf @scopatz @shivamvats

File: series.tex (hunk 4 above)
Comment (shivamvats) [inline series.tex]: @isuruf How do you suggest I add a benchmark? Should I post a link to a wiki page?

**Comment by moorepants** [comment]
This needs to be updated wrt to master and still needs a review.

File: series.tex [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+>>> from sympy.polys.ring_series import rs_sin
+>>> R, x = ring('x', QQ)
+>>> rs_sin(x**2 + x, x, 5)
+-1/2*x**4 - 1/6*x**3 + x**2 + x
+\end{verbatim}
+
+The function \texttt{rs\_series} makes use of these elementary functions to
+intelligently expand an arbitrary SymPy expression. Currently it only supports
+expansion about 0 and is under active development. Ring Series is several times
+faster than the default implementation with the speed difference increasing
```
Comment (leosartaj) [inline series.tex]: You could add a comment here. When you have the benchmarks you can add them here. @shivamvats

**Comment by shivamvats** [comment]
@scopatz Could you please review this?

File: series.tex [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+The first approach stores a series as a core object. Each function has its
+specific implementation of its expansion which is able to evaluate the Puiseux
+series expansion about any point.
+
+\begin{verbatim}
+>>> from sympy import symbols, series
+>>> x, y = symbols('x, y')
+>>> series(sin(x+y) + cos(x*y), x, 0, 2)
+1 + sin(y) + x*cos(y) + O(x**2)
+\end{verbatim}
```
Comment (scopatz) [inline series.tex:19]: I don't think this code block is tied well to the text around it.

File: series.tex [hunk 7]
```diff
@@ -0,0 +1,79 @@
+% Series expansion (Differentiate between the two approaches being used)
+\subsubsection{Series Expansion}
+
+SymPy is able to calculate the symbolic series expansion of an arbitrary series
+or expression involving elementary and special functions and multiple
+variables. For this it has two different implementations.
```
Comment (scopatz) [inline series.tex]: mention the two implementations by name here, so we know what is coming.

File: series.tex [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+The first approach stores a series as a core object. Each function has its
```
Comment (scopatz) [inline series.tex]: What do you mean by `core object` object precisely?

File: series.tex [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+specific implementation of its expansion which is able to evaluate the Puiseux
+series expansion about any point.
```
Comment (scopatz) [inline series.tex]: Citation needed.

File: series.tex [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+\begin{verbatim}
+>>> from sympy import symbols, series
+>>> x, y = symbols('x, y')
+>>> series(sin(x+y) + cos(x*y), x, 0, 2)
+1 + sin(y) + x*cos(y) + O(x**2)
+\end{verbatim}
+
+The newer and much faster approach called Ring Series makes use of the
```
Comment (scopatz) [inline series.tex]: citation needed.

Comment (scopatz) [inline series.tex]: also, I am not sure that the adjective `newer` is useful here. Maybe `alternative`.

Comment (scopatz) [inline series.tex]: Similarly with `faster`, do we have a citation for this or is this something that we can easily prove? If not, we may want to leave it out.

File: series.tex [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+implemented using semi-numerical algorithms. All these support Puiseux series
+expansion.
+
+\begin{verbatim}
+>>> from sympy import ring
+>>> from sympy.polys.ring_series import rs_sin
+>>> R, x = ring('x', QQ)
+>>> rs_sin(x**2 + x, x, 5)
+-1/2*x**4 - 1/6*x**3 + x**2 + x
+\end{verbatim}
```
Comment (scopatz) [inline series.tex:42]: Again please tie this example into the surrounding text. E.g. "The following example demonstrates..."

File: series.tex [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+The function \texttt{rs\_series} makes use of these elementary functions to
```
Comment (scopatz) [inline series.tex]: Should we have the fully qualified name for `rs_series`? Also, should this have parentheses at the end?

File: series.tex [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+intelligently expand an arbitrary SymPy expression. Currently it only supports
```
Comment (scopatz) [inline series.tex]: What do you mean by `intelligently`? This is a loaded term in my opinion :).

File: series.tex [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+faster than the default implementation with the speed difference increasing
+with the size of the series
+
+\begin{verbatim}
+>>> from sympy.polys.ring_series import rs_series
+>>> from sympy.abc import a, b
+>>> from sympy import sin, cos
+>>> rs_series(sin(a + b), a, 4)
+-1/2*(sin(b))*a**2 + (sin(b)) - 1/6*(cos(b))*a**3 + (cos(b))*a
+\end{verbatim}
```
Comment (scopatz) [inline series.tex:60]: textual tie in.

File: series.tex [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+>>> R, x = ring('x', QQ)
+>>> rs_sin(x**2 + x, x, 5)
+-1/2*x**4 - 1/6*x**3 + x**2 + x
+\end{verbatim}
+
+The function \texttt{rs\_series} makes use of these elementary functions to
+intelligently expand an arbitrary SymPy expression. Currently it only supports
+expansion about 0 and is under active development. Ring Series is several times
+faster than the default implementation with the speed difference increasing
+with the size of the series
```
Comment (scopatz) [inline series.tex]: Lack of punctuation.

File: series.tex [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+The advantage of this approach is that an explicit formula for the coefficients
+of the series expansion is generated rather than just computing a few terms.
+
+\begin{verbatim}
+>>> f = fps(sin(x), x, x0=0)
+>>> f.truncate(6)
+x - x**3/6 + x**5/120 + O(x**6)
+>>> f[15]
+-x**15/1307674368000
+\end{verbatim}
```
Comment (scopatz) [inline series.tex:78]: textual tie in.

File: series.tex [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+SymPy provides functionality to compute Fourier Series of a function using
+the \texttt{fourier\_series} function. Under the hood it just computes $a0$, $an$, $bn$ using
+standard integration formulas.
+
+\begin{verbatim}
+>>> L = symbols('L')
+>>> f = fourier_series(2 * (Heaviside(x/L) - Heaviside(x/L - 1)) - 1, (x, 0, 2*L))
+>>> f.truncate(3)
+4*sin(pi*x/L)/pi + 4*sin(3*pi*x/L)/(3*pi) + 4*sin(5*pi*x/L)/(5*pi)
+\end{verbatim}
```
Comment (scopatz) [inline series.tex:93]: Textual tie in.

File: series.tex [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+>>> f = fps(sin(x), x, x0=0)
+>>> f.truncate(6)
+x - x**3/6 + x**5/120 + O(x**6)
+>>> f[15]
+-x**15/1307674368000
+\end{verbatim}
+
+\subsubsection{Fourier Series}
+
+SymPy provides functionality to compute Fourier Series of a function using
```
Comment (scopatz) [inline series.tex]: Citation needed.

File: series.tex (hunk 10 above)
Comment (shivamvats) [inline series.tex]: There are a few speed comparisons [here](https://github.com/sympy/sympy/blob/master/doc/src/modules/polys/ringseries.rst). What sort of citation would do? 

Every series that ring series expands, it does so a couple of times faster than the default series method. We were planning to benchmark the two methods. What would be a good way to go about that?

Comment (scopatz) [inline series.tex]: Maybe you could consolidate the information in that link into a table?

File: series.tex (hunk 6 above)
Comment (shivamvats) [inline series.tex:19]: Resolved.

File: series.tex (hunk 11 above)
Comment (shivamvats) [inline series.tex:42]: Resolved.

File: series.tex (hunk 14 above)
Comment (shivamvats) [inline series.tex:60]: Resolved.

File: series.tex (hunk 16 above)
Comment (shivamvats) [inline series.tex:78]: Resolved.

File: series.tex (hunk 17 above)
Comment (shivamvats) [inline series.tex:93]: Resolved.

File: series.tex [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+consider a Taylor expansion about 0:
+
+\begin{verbatim}
+>>> from sympy import symbols, series
+>>> x, y = symbols('x, y')
+>>> series(sin(x+y) + cos(x*y), x, 0, 2)
+1 + sin(y) + x*cos(y) + O(x**2)
+\end{verbatim}
+
+The newer and much faster\cite{sympyRingSeries} approach called Ring Series makes use of the
```
Comment (shivamvats) [inline series.tex:21]: @scopatz I have cited the SymPy doc here. Will update the doc with more speed comparisons.

**Comment by leosartaj** [comment]
Ping @asmeurer @scopatz

File: series.tex (hunk 14 above)
Comment (scopatz) [inline series.tex:60]: I think there needs to be more than, "An example:"

**Comment by scopatz** [comment]
Thanks @leosartaj! Just the one small issue that I am sure to notice in the full read through.

**Comment by scopatz** [comment]
Thanks to @asmeurer for merging.

**Comment by leosartaj** [comment]
Thanks @asmeurer @scopatz. We will resolve the issue, in the next PR. :smile:

**Pull request merged**